### ISSUE TITLE
Reproducible builds

### ISSUE THREAD

**Issue by IzzySoft** [anonfaded/FadCam#62]
The current release fails RB. The reason is easy: you did not build from a clean tree at the commit the release tag points to, but modified the `app/build.gradle.kts` **after** that commit and then built **without** committing it (and pointing the tag to _that_ commit).

Please always build your release APKs from a clean tree at the commit indicated by the tag. This time, it was easy enough to fix – but that's not always the case. Thanks in advance!

PS: We've prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) you might wish to take a look at, to avoid issues like this :wink:

**Comment by IzzySoft** [comment]
Oh, besides: today's update was rejected by our repo, because…

```
2024-11-21 19:45:09,414 WARNING: "com.fadcam_8.apk" is signed by a key that is not allowed:
a85ed0677555011871e48b5a083a2c5c54464a381c665a068e28d7647deea3bb
```

What happened to your signing key? Your release notes do not mention it. Whoever wants to update to that version will fail, as Android will reject the update as well. Updates at IzzyOnDroid will be disabled until clarified – same will happen at F-Droid as your app there is also pinned to your previous signing key.

Previous key:

```
Signer #1 certificate DN: C=44, ST=UK, L=london, O=fadedhood, OU=fadcam, CN=fadcam
Signer #1 certificate SHA-256 digest: 999e27d8ee66a67319db6827e0f8549c37624601b86f0108c6970413279ef1d4
Signer #1 certificate SHA-1 digest: 323f4b552e345e75fae7927997f8e52a00e99c68
Signer #1 certificate MD5 digest: ab583ae95e9c9a6911b9d365ae6b2e4d
Signer #1 key algorithm: RSA
Signer #1 key size (bits): 2048
```

New key:

```
Signer #1 certificate DN: C=XX, ST=Unknown, L=Unknown, O=FadSec Lab, OU=App Development, CN=Faded Hood
Signer #1 certificate SHA-256 digest: a85ed0677555011871e48b5a083a2c5c54464a381c665a068e28d7647deea3bb
Signer #1 certificate SHA-1 digest: 2cc1eafac5a12fefccf3a5673eb193fa5e17e482
Signer #1 certificate MD5 digest: 6656435a82df413cdb6b68417a911ad3
Signer #1 key algorithm: RSA
Signer #1 key size (bits): 2048
```

I assume you tried to update the DN – which results in a different key. I hope you've kept a copy of the original key?

**Comment by anonfaded** [comment]
> The current release fails RB. The reason is easy: you did not build from a clean tree at the commit the release tag points to, but modified the `app/build.gradle.kts` **after** that commit and then built **without** committing it (and pointing the tag to _that_ commit).
> 
> Please always build your release APKs from a clean tree at the commit indicated by the tag. This time, it was easy enough to fix – but that's not always the case. Thanks in advance!
> 
> PS: We've prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) you might wish to take a look at, to avoid issues like this 😉

I build the apk from the latest master branch idk why this happened but thanks i will read that guide.




> What happened to your signing key?

And yeah i was thinking to discuss this with some staff of F-Droid but good that you asked. Actually i have that key file but i forgot the password as i made this app in July and i didn't know about keys so i randomly just made it at that time. But now i changed my PC and made a new key for this release as i don't have any other option, so please tell me now what to do? Then key file is with me but during generating the apk in android studio, it asks for password and i don't remember that sadly... :')

**Comment by anonfaded** [comment]
oh again i forgot something!
I changed the versionCode and version name locally but i forgot to push that and build with that instead of pushing first, sad.
But now i am pushing it, so please let me know how to manage ahead? Like as i forgot the password for previous key (that i have now too but can't use it without pass), so should i sign with the latest new key  and generate apk and update the ones in tag on github?
I think there is no way for previous key to use without password, but if there is then let me know.

**Comment by IzzySoft** [comment]
As for the signing key: without credentials it's useless. But before you delete it (you might remember the credentials suddenly), please read: [How to keep your key safe and what measures to take for the event of loss?](https://f-droid.org/2023/09/03/reproducible-builds-signing-keys-and-binary-repos.html#lessons-learned-2-how-to-keep-your-key-safe-and-what-measures-to-take-for-the-event-of-loss)

As signing is a security task, we'll now need to verify it's not just some "imposter" having gained access to your repo here. You'll find background and details in the linked article, but in short, what can be taken as verification:

* signed commits: you didn't sign yours (though you let Github sign some, which the imposter could do as well)
* signing with the original key: you can't as you forgot your password
* another main contributor who knows you personally vouching for you: no other "main" contributor as far as I can see
* an alternative contact established before the incident: non established beforehand
* signing with another key that was already known to be yours before: I don't see such an app here at your Github account

So it seems we're in trouble here – unless you see something I might have missed.

> I changed the versionCode and version name locally but i forgot to push that

Yepp, that you did. And that's how I fixed the build here as well, by (automatically) applying the very same change during the build process. Which is what made it RB again – though we cannot show it due to the signing issue. On both ends, F-Droids and ours.

**Comment by anonfaded** [comment]
Wait now everything got fixed😅

I just opened android studio on old PC (good that i didn't uninstall it yet) and then after clicking on 'generate signed apk' i saw the dialog and found the password and name for it. I just realized that i set one name there but now on new pc I was using another name which was causing the error. Now i saved this info and will store it on cloud.

Also i regenerated apks with the original key file so now check again please and now it will pass the build.

Also i want to ask something:
If i commit all things and make tag and upload apks too, but then i do some changes in the fastlane folder lets say i change the metadata, so will this cause issue on F Droid and IoD? Or this isn't linked with apk or tag so it's ok if i do changes and commit to fastlane folder. Thanks!

**Comment by IzzySoft** [comment]
> Wait now everything got fixed😅

Now why was I expecting that? 😅

> I just realized that i set one name there but now on new pc I was using another name which was causing the error.

That's only part of it, AFAIK. Whenever a new key is generated, even with the same DN, I'm not sure if it's identical to the previous one (need to check that to make sure). Better use the original keystore if you can.

The APK currently attached matches your original key. But as it's signed with a different key now than at the time the builder fetched it, the shield won't be up **for this release** – I guess that's something we can live with, the next release should have it up again fine I hope. Much more important the signing got fixed! Now that was achieved, may I strongly recommend to implement one or more of the measures outlined above? Start signing your commits (once established, it's not much overhead; I can help you with the steps to establish it if needed); maybe you have a personal website or (support) email address that could be added to the metadata on our end; and definitely make a good backup of your key as well as put the credentials into some password safe (I can recommend KeePassXC for that, available cross-platform, and as pendant on Android KeePassDX); maybe also name a contact who can vouch for you.

> If i commit all things and make tag and upload apks too, but then i do some changes in the fastlane folder lets say i change the metadata, so will this cause issue on F Droid and IoD?

F-Droid will take the fastlane structure at the tag it builds from. IoD takes it at `HEAD` when the update is fetched (as the updater runs daily, there're usually no differences – but it gives us the flexibility to fetch "forgotten adjustments" easily with a separate call, in case of "urgency").

**Comment by anonfaded** [comment]
Yeah i was using that key from months. But didn't actually knew it was important until i learned about AppStores' requirements lol

And yes guide me on how to sign commits too.

And yeah you can save my info. I can delete the comment after you reply so leaving details below:

**Comment by IzzySoft** [comment]
> And yes guide me on how to sign commits too.

You'll need a GPG/PGP key for that. I don't know what OS you're working on, but there are good articles about setting up GPG and creating your own key. You can have multiple keys, e.g. one for development and another one for personal use.

Next, you'd need to tell your local Git installation which key it should use. You can do that globally (in `$HOME/.gitconfig`), but also per project. This would look like:

```ini
[user]
    email = john.doe@example.com
    name = John Doe
    signingkey = 6752AA5CB1A4521F
```

Now, on each commit you use the `-S` parameter to tell git to sign it. Or you do it like me, and define an alias so you cannot forget:

```ini
[alias]
    ci = commit -S
```

Now, instead of "git commit" you can simply use "git ci".

> And yeah you can save my info. I can delete the comment after you reply so leaving details below:

I'd add those details to the metadata of your app, so they'd be "public" as soon as metadata will be made available via a public git repo (which is planned). So if you need to keep those details secret, that would not work. So which of those details can be made public?

**Comment by IzzySoft** [comment]
PS: I've just [added a page to our wiki](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Developer-Best-Practices/Sign-your-commits) for the commit signing stuff, including some steps for GPG setup.

**Comment by linsui** [comment]
Ah, the tag is on a wrong commit and the apk name format also changed. And the key is also changed? :facepalm:

**Comment by anonfaded** [comment]
> > And yes guide me on how to sign commits too.
> 
> You'll need a GPG/PGP key for that. I don't know what OS you're working on, but there are good articles about setting up GPG and creating your own key. You can have multiple keys, e.g. one for development and another one for personal use.
> 
> Next, you'd need to tell your local Git installation which key it should use. You can do that globally (in `$HOME/.gitconfig`), but also per project. This would look like:
> 
> ```ini
> [user]
>     email = john.doe@example.com
>     name = John Doe
>     signingkey = 6752AA5CB1A4521F
> ```
> 
> Now, on each commit you use the `-S` parameter to tell git to sign it. Or you do it like me, and define an alias so you cannot forget:
> 
> ```ini
> [alias]
>     ci = commit -S
> ```
> 
> Now, instead of "git commit" you can simply use "git ci".
> 
> > And yeah you can save my info. I can delete the comment after you reply so leaving details below:
> 
> I'd add those details to the metadata of your app, so they'd be "public" as soon as metadata will be made available via a public git repo (which is planned). So if you need to keep those details secret, that would not work. So which of those details can be made public?

Hmm I'll look into it, but if i don't do the signing gpg key part it will still be fine i hope?

And you can add that details, they are public. I was just talking about this comment in this issue, nevermind, you can use that all info.

**Comment by anonfaded** [comment]
> Ah, the tag is on a wrong commit and the apk name format also changed. And the key is also changed? :facepalm: 

Now it's all fixed, there were issues that are discussed above. What else issue you need to discuss?

**Comment by linsui** [comment]
The tag still points to a wrong commit. The apk name should be FadCam_v1.2.0_universal.apk as before. Glad that you get the old key back.

**Comment by anonfaded** [comment]
> The tag still points to a wrong commit. The apk name should be FadCam_v1.2.0_universal.apk as before. Glad that you get the old key back.

Oh like to replace hyphens with underscores? But is that really needed? Need confirmation from @IzzySoft

**Comment by linsui** [comment]
@IzzySoft doesn't need that but F-Droid needs that. :shrug:

**Comment by anonfaded** [comment]
> @IzzySoft doesn't need that but F-Droid needs that. :shrug: 

Are you from f droid? He is linked with them somehow tho

**Comment by linsui** [comment]
@IzzySoft is here for their own repo. I'm here for F-Droid main repo. ;) https://gitlab.com/fdroid/fdroiddata/-/merge_requests/16904

**Comment by anonfaded** [comment]
> @IzzySoft is here for their own repo. I'm here for F-Droid main repo. ;) https://gitlab.com/fdroid/fdroiddata/-/merge_requests/16904

Ok thanks.
So should i just rename the names and replace - with _ ? Is this enough?

**Comment by IzzySoft** [comment]
> Hmm I'll look into it, but if i don't do the signing gpg key part it will still be fine i hope?

If you forget it once in a while, nobody will die of it. It's an additional security measure.

> And you can add that details, they are public. I was just talking about this comment in this issue, nevermind, you can use that all info.

Thanks! Then I'll add them now… done.

> @IzzySoft doesn't need that but F-Droid needs that. 🤷

@linsui is right there. All IzzyOnDroid has in that context is: `ApkMatch: /armeabi-v7a/i` – so that's the part that should be stable and unique for each release. And he's also right that, for this issue, I'm here for IzzyOnDroid – while linsui is asking for the builds at F-Droid.org.

To answer your question directed at him: you should decide for one variant, and then stick with it (so either hyphens or underscores – not one time these and another those). I'd suggest you leave the ones at v1.2.0 as they are (that build would fail at F-Droid anyway as the signing key is wrong), and go back to the original naming conventions with the next release, so F-Droid can pick it up properly.

@linsui guess you want to simply disable that tag/build on your end – it will fail RB as well as signature check. It failed both at IzzyOnDroid, too – you know we also pin signing keys. The next release will hopefully be fine again.

**Comment by anonfaded** [comment]
> you should decide for one variant, and then stick with it (so either hyphens or underscores – not one time these and another those).

Okay, from next releases i will stick to hyphens only. Also next release will be pushed very soon so if this one has issues then IoD and F-Droid can wait for next update.

**Comment by linsui** [comment]
> guess you want to simply disable that tag/build on your end – it will fail RB as well as signature check. It failed both at IzzyOnDroid, too – you know we also pin signing keys. The next release will hopefully be fine again.

I thought that the apk has been updated since the old password is found...

Then I'll wait for the next release, thanks!

**Comment by IzzySoft** [comment]
> I thought that the apk has been updated since the old password is found...

Ah, right, the additional issue of replacing the APKs which had already been distributed – apologies, I got a bit lost here :see_no_evil: So indeed, with the file names adjusted, they could work at F-Droid.

Well, IoD would not process that tag again. But rewriting/replacing an already distributed release is not exactly "best practice"; others might have picked up things as well and get confused (and RB results are becoming weird if the referenced file name disappeared). So if the next release is already close, as @anonfaded just stated, maybe best is waiting for that. Your choice of course :wink:

**Comment by anonfaded** [comment]
> > I thought that the apk has been updated since the old password is found...
> 
> Ah, right, the additional issue of replacing the APKs which had already been distributed – apologies, I got a bit lost here 🙈 So indeed, with the file names adjusted, they could work at F-Droid.
> 
> Well, IoD would not process that tag again. But rewriting/replacing an already distributed release is not exactly "best practice"; others might have picked up things as well and get confused (and RB results are becoming weird if the referenced file name disappeared). So if the next release is already close, as @anonfaded just stated, maybe best is waiting for that. Your choice of course 😉

Yeah next release will be probably in 2-3 days, Thanks!

**Comment by anonfaded** [comment]
> Days, or weeks? 🤔

Actually I'm waiting for a contributor who is fixing some issues, there is some high priority issue with video codec, so as soon as he is done, i will let you know.

Btw some days ago someone told me they downloaded the FadCam app from IzzyOnDroid and it wasn't working on samsung s24, due to armeabi build. So, using the arm64 next time is better i think.

**Comment by IzzySoft** [comment]
> there is some high priority issue with video codec, so as soon as he is done, i will let you know.

Thanks – and best luck with that issue then!

> using the arm64 next time is better i think.

Due to the size of the APKs, we can only have either-or. We can switch to arm64, but then people with 32bit devices may complain. To my knowledge there are rather few 64bit devices not supporting 32bit APKs (this is only the 2nd device I know of now), but no 32bit device (and yes, there are still some around) can run arm64 APKs. At IzzyOnDroid, we favor longevity and sustainability (i.e. using your (older) device as long as possible instead of producing more e-waste by switching every 2nd year) over having "the latest and greatest", hence we usually pick armeabi in such cases. Of course, as author you have a say there, too; if you say you want it that way, we can switch to arm64 here with the next release.

**Comment by anonfaded** [comment]
@IzzySoft @linsui 
Hi, finally new version is released and now we can proceed with IOD and F-Droid to update the app there with updated `fastlane` metadata.
https://github.com/anonfaded/FadCam/releases/tag/v1.3.0

**Comment by IzzySoft** [comment]
Could it be you've switched to JDK-21 for this release? It's no longer RB with 17, and the DEX diff looks like "17 vs 21" (additional annotations). Et voila, with JDK-21 it's RB again.

@linsui ^^ might save you a build cycle 😉

**Comment by anonfaded** [comment]
> Could it be you've switched to JDK-21 for this release? It's no longer RB with 17, and the DEX diff looks like "17 vs 21" (additional annotations). Et voila, with JDK-21 it's RB again.
> 
> @linsui ^^ might save you a build cycle 😉 

I did a lot of refactoring and i guess yes.
So is this again an issue? This time i want to must update the F-Droid version as already its 8 month old there 😅

**Comment by IzzySoft** [comment]
No, it's not an issue – we just need to adjust our recipes accordingly. For the builders at IzzyOnDroid, this has already been done – I've explicitly mentioned it here so linsui can adjust that for F-Droid, to avoid further delay.

Btw, as you use shields on your Readme, maybe you want to add [one for RB](https://shields.rbtlog.dev/) as well? 😉

**Comment by anonfaded** [comment]
Oh fine, finally a release without any issue haha.

And i added `![](https://shields.rbtlog.dev/simple/com.fadcam)` in the readME.

**Comment by linsui** [comment]
@IzzySoft Thanks! :)

**Comment by linsui** [comment]
Oh, these libs need to be built from source. https://github.com/anonfaded/FadCam/tree/master/app/libs And some patches are required for reproducible build. See https://github.com/InfinityLoop1308/ffmpeg-kit/commits/main/.

**Comment by anonfaded** [comment]
> Oh, these libs need to be built from source. https://github.com/anonfaded/FadCam/tree/master/app/libs And some patches are required for reproducible build. See https://github.com/InfinityLoop1308/ffmpeg-kit/commits/main/.

Can you guide what I'm supposed to do about it please? I didn't understand, i added everything already as part of FadCam and it builds fine tho

**Comment by linsui** [comment]
You need to build the ffmpeg-kit from source with the patches for reproducible build so that we can also build the same aar. You can just use https://github.com/InfinityLoop1308/PipePipeClient/blob/dev/ffmpeg/ffmpeg-kit.aar which we have built for PipePipe.

**Comment by anonfaded** [comment]
> You need to build the ffmpeg-kit from source with the patches for reproducible build so that we can also build the same aar. You can just use https://github.com/InfinityLoop1308/PipePipeClient/blob/dev/ffmpeg/ffmpeg-kit.aar which we have built for PipePipe.

so should i just download this aar and replace it with mine? The name should be `ffmpeg-kit.aar` and not what i have in mine one? Also is this the `full` version which i m using in my app?
And after i replace the aar, will you be able to use this existing tag or should i delete this tag and redo the release with a new tag?

**Comment by linsui** [comment]
It's not the full version. It's built with `./android.sh --lts --enable-gmp --enable-openssl --enable-android-media-codec --enable-android-zlib --disable-x86 --disable-arm-v7a --disable-arm-v7a-neon`. If you need a full version you can build one with those patches. Then we need to build a same aar. Gen You need to create a new release for that.

**Comment by anonfaded** [comment]
> you can build one with those patches.

But this is what i dont know how to do, can you guide please. I will delete the v1.3.0 tag and will make it again when done but for now i want to fix this issue.
I also have this fork if this helps:
https://github.com/anonfaded/ffmpeg-kit

**Comment by linsui** [comment]
Basically you need to install deps then run `ANDROID_SDK_ROOT=... ANDROID_NDK_ROOT=... ./android.sh` with args. If you want the full version then you need to pass `--full` and `--enable-gpl` to it. But it will take a long time to build and I'm not sure if all libs are reproducible. We need to check that.

**Comment by IzzySoft** [comment]
> And i added `![](https://shields.rbtlog.dev/simple/com.fadcam)` in the readME.

You can link it to `https://shields.rbtlog.dev/com.fadcam` if you want (that's where it shows some details) 😉

**Comment by anonfaded** [comment]
> You can link it to `https://shields.rbtlog.dev/com.fadcam` if you want (that's where it shows some details) 😉

i used `![](https://shields.rbtlog.dev/com.fadcam)` but it didn't show any badge. Is it correct?

**Comment by IzzySoft** [comment]
`[<img src="https://shields.rbtlog.dev/simple/com.fadcam" alt="badge">](https://shields.rbtlog.dev/com.fadcam)`

[<img src="https://shields.rbtlog.dev/simple/com.fadcam" alt="badge">](https://shields.rbtlog.dev/com.fadcam)

**Comment by anonfaded** [comment]
> `[<img src="https://shields.rbtlog.dev/simple/com.fadcam" alt="badge">](https://shields.rbtlog.dev/com.fadcam)`
> 
> [![badge](https://camo.githubusercontent.com/2ea5c11870091912950a7dcf1582acd2d71e0c6eb64ab3708dca22b4cdb2483a/68747470733a2f2f736869656c64732e7262746c6f672e6465762f73696d706c652f636f6d2e66616463616d)](https://shields.rbtlog.dev/com.fadcam)

It is exact same tho:
![Image](https://github.com/user-attachments/assets/ce6e1056-4440-41b5-85ee-4f847f528b4a)

**Comment by anonfaded** [comment]
> Basically you need to install deps then run `ANDROID_SDK_ROOT=... ANDROID_NDK_ROOT=... ./android.sh` with args. If you want the full version then you need to pass `--full` and `--enable-gpl` to it. But it will take a long time to build and I'm not sure if all libs are reproducible. We need to check that.

I'm trying `./android.sh --full --enable-gpl --api-level=28`, will let you know if it works

**Comment by IzzySoft** [comment]
> It is exact same tho:

It **looks** the same, yes. But other than now, it would link to details – instead to some copy of the very same image at camo.githubusercontent.com 😉 Which makes it more usable, as visitors can directly check what it means.

**Comment by IzzySoft** [comment]
Eh, what's that now?

```
WARNING: "com.fadcam_11.apk" is signed by a key that is not allowed:
a85ed0677555011871e48b5a083a2c5c54464a381c665a068e28d7647deea3bb
```

Not again, please. That's the very same "wrong key" you had last year in November.

Next, the APK size is WHAT? 60 MB for the armeabi build? What happened? That's almost 3 times the size of the last one we had. And twice the per-app size limit here. So we had to disable updates on our end again. We had verified your (new) key back then, but according to my notes switched back to the restored original one (the comments above seem to confirm that, if I read them correctly).

So what are we doing now? Which key will you be using (remember, changing the key forces everyone who already uses your app to uninstall and re-install, as direct updates are not possible without proper key rotation – which is supported at IzzyOnDroid but no longer at F-Droid)? And what will we do about the APK size? We could switch to the arm64 build, but even that is already beyond our 30MB margin (though rather slightly).

I see other than the arm64 build, armeabi has 2 `libavcodec` libraries with ~13 MB each, `libavcodec.so` and `libavcodec_neon.so`. Same for `libavfilter` and `libavformat` with ~ 5 MB each – that explains the difference in APK size, as those libraries are stored uncompressed. The arm64 ABI also grew by ~15 MB compared to the previous release; I guess that has something to do with the ffmpeg build? You've built the full variant now, and thus blew the size?

**TL;DR:** that update was rejected due to the wrong signing key, and further updates are currently disabled until signing & size problems are solved, sorry…

**Comment by anonfaded** [comment]
> That's the very same "wrong key"

Ah i was mistaken, i forgot to check the dates of both keys that i have, in the 'old' folder i have the original key and i thought i had to use the one out of folder. But i just checked the dates and i by mistake used the new key again. 
![Image](https://github.com/user-attachments/assets/a4d76adc-fdc9-499f-b175-4c78672c303c)

But now i just updated and used the correct key, so just for testing uploaded this apk which you can examine for key confirmation if you want. But the main issue is now the F-Droid which i want to solve and after that i will be able to release a new version which then you can use for IOD.
apk link for testing (if you want): https://github.com/anonfaded/FadCam/releases/download/debug-60fps/FFmpeg-arm64.apk

And about the apk size, i tried but i can't decrease it anymore as it has important libs which can't be removed. Right now i am using a custom `aar` for ffmpeg which i built with this command with these libs included:

```bash
./android.sh --disable-x86 --disable-x86-64 --disable-arm-v7a-neon --enable-openssl --enable-freetype --enable-lame --enable-libiconv --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libwebp --enable-opus --enable-soxr --enable-libpng --enable-jpeg --enable-zimg --enable-x264 --enable-gpl --enable-android-media-codec --api-level=28
```

And the armeabi apk right now is 37MB and arm64 is 39MB

**Comment by anonfaded** [comment]
> Basically you need to install deps then run `ANDROID_SDK_ROOT=... ANDROID_NDK_ROOT=... ./android.sh` with args. If you want the full version then you need to pass `--full` and `--enable-gpl` to it. But it will take a long time to build and I'm not sure if all libs are reproducible. We need to check that.

@linsui I finally built the aar so sharing the details below.

I used `android-ndk-r25c` NDK and instead of full, i made it with these libs so check the command:

```bash
./android.sh --disable-x86 --disable-x86-64 --disable-arm-v7a-neon --enable-openssl --enable-freetype --enable-lame --enable-libiconv --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libwebp --enable-opus --enable-soxr --enable-libpng --enable-jpeg --enable-zimg --enable-x264 --enable-gpl --enable-android-media-codec --api-level=28
```
- And the aar that i built is available here:
https://github.com/anonfaded/FadCam/commit/66d5c863bbe86d01ee667c924017523023795d38

- The ffmpeg-kit repo that i used:
https://github.com/anonfaded/ffmpeg-kit

So now please let me know is it now able to pass everything for F-Droid? And what are the next steps? I'll wait for your response. Once everything is done then i will release a new version of app which then you can pick for F-Droid.
The wrong keystore issue is also solved now.

**Comment by IzzySoft** [comment]
> i can't decrease it anymore as it has important libs which can't be removed.

I suspect the ffmpeg changes. If I read the above correctly, you now switched to a "fuller" variant you didn't have before. Might be an option to offer a "light" build variant – without the additional ffmpeg components.

> `--disable-arm-v7a-neon`

Funny. Was that there from the beginning, or just added recently? Because last time I checked, the neon stuff was in the armeabi APK.

> And the armeabi apk right now is 37MB and arm64 is 39MB

Ah, then it must have been after, OK. Well, 20% over the size limit, guess we can close an eye as long as it doesn't go beyond 40 MB then. Please give me a ping once the next release is out, so I check and re-enable updates (which were disabled because of the key and size issues, which are both addressed now).

**Comment by anonfaded** [comment]
> Might be an option to offer a "light" build variant – without the additional ffmpeg components.

Already tried the lighter versions, always had errors, hence i changed to full version and that was only working. 

> Was that there from the beginning, or just added recently?

As this time the issue is with the ffmpeg and F-Droid, so i had to go thru documentation and build it myself. So i just disabled it as its useless for my app. 
Also i asked Ai for what libs to include for audio and video purposes in my app so i don't have to rebuild from scratch later when i add video editing feature in the app, so it suggested the libs that i mentioned in the `./android.sh` command above.



> we can close an eye as long as it doesn't go beyond 40 MB

Thank you! Yeah it won't go above this hopefully as now there aren't needed any major changes that involves more libs.



> Please give me a ping once the next release is out

Sure, I'm awaiting response from F-Droid staff once they confirm that now the pipeline works for them, i'll release next version.

**Comment by linsui** [comment]
Which jdk do you use to build the ffmpeg-kit? What's the local path of the ffmpeg-kit repo when you build it? And these patches are required to make the build reproducible. https://github.com/InfinityLoop1308/ffmpeg-kit/compare/86b08f0701352a7daaf6cec72625d7b6876dfae4...InfinityLoop1308:ffmpeg-kit:efd648489598efdc65e2e2f06ae69e065e257077

**Comment by anonfaded** [comment]
> Which jdk do you use to build the ffmpeg-kit?

i used this:
```bash
openjdk version "17.0.15" 2025-04-15
OpenJDK Runtime Environment (build 17.0.15+6-Ubuntu-0ubuntu124.04)
OpenJDK 64-Bit Server VM (build 17.0.15+6-Ubuntu-0ubuntu124.04, mixed mode, sharing
```


> What's the local path of the ffmpeg-kit repo when you build it?

I used Ubuntu WSL, this is the path `\\wsl.localhost\Ubuntu-24.04\home\faded\ffmpeg-kit`


> And these patches are required to make the build reproducible.

I updated my fork with these patches, now what to do?
https://github.com/anonfaded/ffmpeg-kit/commit/21bd5bf2a54d71d551d653c7925975e47847fddc

**Comment by linsui** [comment]
> I used Ubuntu WSL, this is the path \\wsl.localhost\Ubuntu-24.04\home\faded\ffmpeg-kit

I'm not sure if this will cause any problem. We need to test it. Can you upload the aar built from https://github.com/anonfaded/ffmpeg-kit/commit/21bd5bf2a54d71d551d653c7925975e47847fddc?

**Comment by anonfaded** [comment]
> > I used Ubuntu WSL, this is the path \wsl.localhost\Ubuntu-24.04\home\faded\ffmpeg-kit
> 
> I'm not sure if this will cause any problem. We need to test it. Can you upload the aar built from [anonfaded/ffmpeg-kit@21bd5bf](https://github.com/anonfaded/ffmpeg-kit/commit/21bd5bf2a54d71d551d653c7925975e47847fddc)?

It is currently in progress. Will ping you when it is done

**Comment by anonfaded** [comment]
> > I used Ubuntu WSL, this is the path \wsl.localhost\Ubuntu-24.04\home\faded\ffmpeg-kit
> 
> I'm not sure if this will cause any problem. We need to test it. Can you upload the aar built from [anonfaded/ffmpeg-kit@21bd5bf](https://github.com/anonfaded/ffmpeg-kit/commit/21bd5bf2a54d71d551d653c7925975e47847fddc)?

@linsui 
I uploaded the aar here https://github.com/anonfaded/FadCam/commit/e2c56e953b4bd3240c9a11738390395ae1f726f6
Please check and let me know how it goes

**Comment by linsui** [comment]
Great, the first try builds an identical aar. But there is a problem that installing jdk 21 messed up the packages. Can you switch back the jdk 17 for app build? Can you remove the jar files in https://github.com/anonfaded/FadCam/tree/master/app/libs and use `implementation("com.arthenica:smart-exception-java:0.2.1")` instead?

**Comment by linsui** [comment]
I solved the jdk 21 problem. So please remove the jars and release a new version. :)

**Comment by anonfaded** [comment]
> I solved the jdk 21 problem. So please remove the jars and release a new version. :)

Thank you so much! I removed the jars https://github.com/anonfaded/FadCam/commit/b9462fdb0d7bdd758becc068d9dabe8d1a713f54
and will let you know in some hours when i release new version.

**Comment by IzzySoft** [comment]
> I solved the jdk 21 problem. So please remove the jars and release a new version. :)

Anything I should keep in mind for our RBs then, @linsui ?

**Comment by linsui** [comment]
I thought your build won't be affected. @IzzySoft

**Comment by anonfaded** [comment]
@linsui @IzzySoft 
New version is released.
https://github.com/anonfaded/FadCam/releases/tag/v1.4.0

**Comment by IzzySoft** [comment]
> I thought your build won't be affected.

We'll get to know soon, @linsui – as @anonfaded just made a new release. But if the build recipe changed (from using an `*.aar` to creating it first), that might have effect. You had some problem to solve concerning JDK-21, so I thought it might show up here as well…

**Comment by IzzySoft** [comment]
Ah, that one was built on Windows. OK, small fix, RB now.

**Comment by linsui** [comment]
Interesting. WSL also uses CRLF?

**Comment by IzzySoft** [comment]
Looks like. Had them in `META-INF/services/*` (the usual suspect for that). I was already suspecting something strange seeing the path @anonfaded posted had Windows UNC notation (backward slashes). I've never worked with WSL, so I cannot tell if that's the way it's supposed to be.

**Comment by anonfaded** [comment]
> Looks like. Had them in `META-INF/services/*` (the usual suspect for that). I was already suspecting something strange seeing the path @anonfaded posted had Windows UNC notation (backward slashes). I've never worked with WSL, so I cannot tell if that's the way it's supposed to be.

I actually opened the file explorer of windows and from there opened the location of ubuntu and then copied it from the address bar.
I didn't copy from inside the terminal. In terminal it is normal like `home/user/ffmpeg-kit/*`, that caused confusion i guess...

**Comment by IzzySoft** [comment]
Indeed. But that still leaves the question raised by @linsui: we'd not expect a Linux build coming with Windows line endings, But somehow, yours did. Other than copying the path, did you run the build inside Linux, or also "from the outside"?

**Comment by anonfaded** [comment]
> Indeed. But that still leaves the question raised by @linsui: we'd not expect a Linux build coming with Windows line endings, But somehow, yours did. Other than copying the path, did you run the build inside Linux, or also "from the outside"?

I did inside the wsl terminal where it's all linux, idk why it used crlf instead of LF.

**Comment by linsui** [comment]
Anyway, this issue can be closed. :)

**Comment by anonfaded** [comment]
> Anyway, this issue can be closed. :)

I'll close it. Btw when will the app update on F-Droid?

**Comment by linsui** [comment]
It should be updated in a week.

**Comment by IzzySoft** [comment]
> I did inside the wsl terminal where it's all linux, idk why it used crlf instead of LF.

Thanks! Just wanted to make sure. Always good to know what's going on – even if we cannot always tell why it does.

> Btw when will the app update on F-Droid?

It should be live at IzzyOnDroid with the next sync. By the way, our updater reported:

    ! repo/com.fadcam_12.apk declares sensitive permission(s):
      android.permission.BLUETOOTH_SCAN android.permission.ACCESS_MEDIA_LOCATION

Could you please clarify what these two are used/needed for?

**Comment by anonfaded** [comment]
I added two new features in app.
`android.permission.BLUETOOTH_SCAN` is used for this following option where user can select different mic as audio source to use in recording
![Image](https://github.com/user-attachments/assets/f5fb9919-7f5b-4996-8d28-6595f87f2b56)

and `android.permission.ACCESS_MEDIA_LOCATION` is used to read metadata from videos that FadCam will save. If the following option is enabled then in the metadata of video, location data can be saved which can then be viewed in any gallery software like google photos for example. This permission is mainly to read the metadata from a file. I'll add a new option soon where this metadata will be shown with the video info so user can easily check the location instead of saving video and using another gallery app just to check what metadata it has saved.
![Image](https://github.com/user-attachments/assets/5a9e9704-1c52-47d3-beb2-30b94d7506ec)

**Comment by IzzySoft** [comment]
Thanks! Added the explanation for Bluetooth here. As for "read metadata from videos that FadCam will save": please correct me if I'm wrong, but if it's just about the videos the app recorded by itself, doesn't scoped storage allowed that by default, as the app is registered as the owner? Just wondering.

**Comment by anonfaded** [comment]
> doesn't scoped storage allowed that by default

I'll test the app without this permission and check if it works, and if it does then in next release i can remove this. Ai just suggested to add this when i was implementing that feature and i did without much research:)

**Comment by anonfaded** [comment]
> doesn't scoped storage allowed that by default

You are right, it works without this permission. I removed the permission and in next releases it will not be there but the current app version; you can sync it i guess

**Comment by IzzySoft** [comment]
> Ai just suggested to add this

Ai also [suggest you eat a stone each day](https://www.reddit.com/r/google/comments/1cziil6/a_rock_a_day_keeps_the_doctor_away/) 🙊 💨 

> You are right, it works without this permission. I removed the permission

Wonderful, thanks!

**Comment by anonfaded** [comment]
> Ai also [suggest you eat a stone each day](https://www.reddit.com/r/google/comments/1cziil6/a_rock_a_day_keeps_the_doctor_away/) 🙊 💨

This is hilarious 🤣
But most of the times I'm in a hurry so i don't cross check haha

**Comment by IzzySoft** [comment]
> But most of the times I'm in a hurry so i don't cross check haha

You always should, as these 2 examples showed. I know it's easy to just get spoon-fed answers, but not even checking their correctness will lead to "brain rot", we'd unlearn to think for ourselves. That in addition to potential wrong decisions – it's not always as obvious as with the rocks (or with [Pizza Glue](https://www.reddit.com/r/ChatGPT/comments/1cz164a/googles_ai_search_feature_suggested_using_glue_to/) at that)…

**Comment by anonfaded** [comment]
> You always should, as these 2 examples showed. I know it's easy to just get spoon-fed answers, but not even checking their correctness will lead to "brain rot", we'd unlearn to think for ourselves.

This is very true, previously this happened a lot with code parts in the app, but now i check the changes before making it final to confirm it didn't alter something else. In this case i just ignored the permissions but yeah i should double check. Learned a lot from you and linsui in these past months, Thanks for all your help!

**Comment by IzzySoft** [comment]
And thanks to you for carrying this through – it was some effort on your end as well! 🤗

Guess we can close this issue now – unless something was left unsolved?

**Comment by anonfaded** [comment]
> Guess we can close this issue now – unless something was left unsolved?

Yeah finally everything is solved, for more issues if any; we can always discuss again here or in a new issue:)

**Comment by IzzySoft** [comment]
> You are right, it works without this permission. I removed the permission and in next releases it will not be there

Well… now instead it is `android.permission.QUERY_ALL_PACKAGES`.

**Comment by anonfaded** [comment]
> Well… now instead it is `android.permission.QUERY_ALL_PACKAGES`.

This permission is added for these new features:

![Image](https://github.com/user-attachments/assets/9e290fb0-b31c-4213-bade-fba6e74dd711)

and these options trigger the intents to open the file upload screens directly in those apps:

![Image](https://github.com/user-attachments/assets/6dc686a2-7c19-4db0-9607-78756b99b060)

i added this for checking if the app is installed and then use the intents but yeah i can directly use intents as well and i'll do that. Currently I'm finalizing the OpenGL setup and doing refactoring so after that i'll come at this.

**Comment by IzzySoft** [comment]
Ah, OK – so "to find compatible apps for uploads to YouTube/Google Drive" should match that? I've set that now, thanks!

**Comment by anonfaded** [comment]
> Ah, OK – so "to find compatible apps for uploads to YouTube/Google Drive" should match that? I've set that now, thanks!

yes correct. this is used to query if youtube and google drive apps are installed on the device and if it is then it opens their file upload screens with the video file loaded.
Should i not remove this permission now?

**Comment by IzzySoft** [comment]
No, why? If it's needed, it's needed. We just feel it important to make transparent what for, so folks can make informed decisions. That way, what might look "suspicious" on a first glance, is explained on a second and avoids "prejudices" etc. So thanks for clarifying!

If you know a different way not requiring such more sensitive permissions, then that should of course be preferred. But if there is not, you can't 😉

**Comment by anonfaded** [comment]
Hi @IzzySoft,
i recently got messages from some users that the IzzyOnDroid repo's fadcam version doesn't work on their phones, and the reason is armeabi ABI. Also a user raised an issue on this same matter: https://github.com/anonfaded/FadCam/issues/137#issuecomment-3001022832
So i think from next releases you should use the arm64 ABI as it is compatible with most phones. The users who complained about this are mostly using S-series Samsung devices

**Comment by IzzySoft** [comment]
You're aware of two things here I hope:

1. that way, support for older 32bit-only devices is dropped
2. the 32bit APK is already beyond our 30 MB per-app size limit. The 64bit version is even larger.

So do you in turn see a possibility to reduce APK size? One option would e.g. be to turn on compression for native libs (which are currently only stored, not compressed, as this is a legacy option).
